File: problems/70/70.climbing-stairs.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#

# @lc code=start
class Solution:
    def climbStairs(self, n: int) -> int:
        # dp[i] = dp[i - 1] + dp[i - 2]
        # dp = [0, 1, 2]

        dp = [1] * (n + 1)

        for i in range(2, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]

        return dp[n]


# @lc code=end

File: problems/70/70.climbing-stairs.java
/*
 * @lc app=leetcode id=70 lang=java
 *
 * [70] Climbing Stairs
 */

// @lc code=start
class Solution {
    public int climbStairs(int n) {
        /*
         * takes n steps to reach top
         * each time can climb either 1 or 2 steps.
         * how many distinct ways can you climb to the top?
         *
         * n = 2
         * 1. 1 + 1
         * 2. 2
         *
         * intuition:
         * to reach step i, can take either 1 or 2 steps
         * dp[i] = dp[i - 1] + dp[i - 2]
         * only prev 2 matter
         *
         * O(N) time, O(1) space
         */

        int prev = 1;
        int curr = 1;

        for (int i = 1; i < n; i++) {
            int temp = curr;
            curr += prev;
            prev = temp;
        }

        return curr;
    }
}
// @lc code=end
